### ISSUE TITLE
"Could not detect shell prompt (timeout)"

### ISSUE THREAD

**Issue by maxentile** [radical-cybertools/radical.pilot#1139]
Hello!

I've been trying to run `examples/00_getting_started.py` locally, but I keep encountering the following error:

```
================================================================================
 Getting Started (RP version 0.43.1)                                            
================================================================================

new session: [rp.session.Joshuas-MacBook-Pro.local.joshuafass.017094.0025]     \
database   : [mongodb://rp:rp@ds015335.mlab.com:15335/rp]                     ok
read config                                                                   ok

--------------------------------------------------------------------------------
submit pilots                                                                   

create pilot manager                                                          ok
create pilot description [local.localhost:2]                                  ok
submit 1 pilot(s) 2016-10-20 16:48:25,077: radical.saga.pty    : MainProcess                     : MainThread     : ERROR   : Could not detect shell prompt (timeout) (/Users/joshuafass/myenv/lib/python2.7/site-packages/saga/utils/pty_shell_factory.py +290 (_initialize_pty)  :  raise se.NoSuccess ("Could not detect shell prompt (timeout)"))
Traceback (most recent call last):
  File "/Users/joshuafass/myenv/lib/python2.7/site-packages/saga/utils/pty_shell_factory.py", line 290, in _initialize_pty
    raise se.NoSuccess ("Could not detect shell prompt (timeout)")
NoSuccess: Could not detect shell prompt (timeout) (/Users/joshuafass/myenv/lib/python2.7/site-packages/saga/utils/pty_shell_factory.py +290 (_initialize_pty)  :  raise se.NoSuccess ("Could not detect shell prompt (timeout)"))
caught Exception: Could not detect shell prompt (timeout) (/Users/joshuafass/myenv/lib/python2.7/site-packages/saga/utils/pty_shell_factory.py +290 (_initialize_pty)  :  raise se.NoSuccess ("Could not detect shell prompt (timeout)"))
```

After taking a look at similar reports on the Issue tracker (and the saga-python Issue tracker), it looks like this could have many possible causes...

So far, I've only tried setting `SAGA_PTY_SSH_TIMEOUT` to a larger number, uninstalling and reinstalling, and restarting. Any ideas what I should try next?

Versions:
- Mac OS 10.12 Sierra
- Python 2.7.11
- MongoDB v3.2.10

Many thanks,
Josh

**Comment by andre-merzky** [comment]
Ah, the dreaded shell-prompt detection... :)

I don't think you did anything wrong with respect to versions and installation.  The problem usually indicates that we are, on a lower level, unable to parse (or to get) the output from an ssh channel.

Would you mind running the example again, but additionally set `RADICAL_SAGA_PTY_VERBOSE=DEBUG`, and to post the output?  Feel free to send it to me by mail (it will be large): andre@merzky.net.

Thanks, Andre.

**Comment by andre-merzky** [comment]
Quick note: the logs show `zsh` as login shell.  The current theory is that we stumble over zsh prompt coloring again.

**Comment by maxentile** [comment]
Thank you so much for the quick response, Andre!

I switched to bash and this issue resolved, and the example `00_getting_started.py` now runs without error.

**Comment by andre-merzky** [comment]
Ah, glad to hear that!  I'll add some notes from the mails to the ticket, and will close then.

We'll add the shell limitations to the FAQ.

**Comment by marksantcroos** [comment]
We might even want to as far as checking for the shell and failing early with a not-supported message.

**Comment by andre-merzky** [comment]
We can't really check easily for remote connections, because we need to get on the machine to do so in the first place.  For local we could do that on startup indeed - and at least issue a big fat warning...

**Comment by andre-merzky** [comment]
For the record: here are the relevant parts from the logs sent by Josh:

```
[...]
2016-10-21 10:28:15,013: radical.saga.pty    : MainProcess                     : MainThread     : INFO    : running: /usr/bin/env TERM=vt100 /bin/zsh -i
[...]
2016-10-21 10:28:17,663: radical.saga.pty    : MainProcess                     : PilotLauncherWorker-1: DEBUG   : read : [   13] [  114] (HELLO_1_SAGA\n%                                                                                )
2016-10-21 10:28:17,663: radical.saga.pty    : MainProcess                     : PilotLauncherWorker-1: DEBUG   : got shell prompt trigger (5) ('
%                                                                                $ printf 'HELLO_%d_SAGA\n' 1
HELLO_1_SAGA)
2016-10-21 10:28:17,666: radical.saga.pty    : MainProcess                     : PilotLauncherWorker-1: DEBUG   : read : [   13] [   89] ()
2016-10-21 10:28:17,666: radical.saga.pty    : MainProcess                     : PilotLauncherWorker-1: DEBUG   : read : [   13] [   32] ($)
2016-10-21 10:28:27,655: radical.saga.pty    : MainProcess                     : PilotLauncherWorker-1: ERROR   : Could not detect shell prompt (timeout) (/Users/joshuafass/myenv/lib/python2.7/site-packages/saga/utils/pty_shell_factory.py +290 (_initialize_pty)  :  raise se.NoSuccess ("Could not detect shell prompt (timeout)"))
[...]
```

**Comment by shantenujha** [comment]
There is a fundamental problem with design if the color of a prompt screws
up downstream functionality.

**Comment by andre-merzky** [comment]
Yes - the fundamental design problem is still that we do screen scraping from an interactive ssh session, which is error prone and volatile.  We stick to it because it is the only way to maintain gsissh support.

At least in Python land, we don't see any viable alternative, as we discussed multiple times.  On C-level, we could code our own gsi and ssl enabled client -- but that is likely an effort over several man-years.  Or we could use the globus libs, which we stopped using (at least directly) some years back, for good reasons.  Or we could switch to a server based architecture, which removes some constraints in this context.  Or we could drop gsissh support, and then use paramiko (which has its own share of troubles though, prompt detection being among them, FWIW, see discussion on fabric/fabric/issues/7).

Either way, as long as we stay in Python and want to support gsissh, those are not really alternatives.
